include/dynd/kernels/string_algorithm_kernels.hpp

```diff
@@ -3,83 +3,33 @@
 // BSD 2-Clause License, see LICENSE.txt
 //
 
-// Implement a number of string algorithms. Note that the initial
+// Implement a number of string algorithms.
 
 #pragma once
 
+#include <dynd/callable.hpp>
 #include <dynd/kernels/kernel_builder.hpp>
-#include <dynd/string_encodings.hpp>
+#include <dynd/types/string_type.hpp>
 
-namespace dynd { namespace kernels {
+namespace dynd {
+  namespace nd {
 
-/**
- * String concatenation kernel.
- *
- * (string, string, ...) -> string
- */
-struct DYND_API string_concatenation_kernel {
-    typedef string_concatenation_kernel extra_type;
+    struct string_concatenation_kernel
+      : base_kernel<string_concatenation_kernel, 2> {
+      typedef string_concatenation_kernel extra_type;
 
-    ckernel_prefix m_base;
-    // The number of input operands
-    size_t m_nop;
+      void single(char *dst, char *const *src);
+    };
 
-    ckernel_prefix& base() {
-        return m_base;
-    }
+  } // namespace nd
 
```

Comment:
Your string_concatenation kernel is perfect. You just need to fill in the body of single with the operation, and that's exactly how we write them.